lib/modules/manager/gomod/artifacts.ts

```diff
@@ -443,37 +456,30 @@ export async function updateArtifacts({
   }
 }
 
-function getGoConstraints(content: string): string | undefined {
-  // prefer toolchain directive when go.mod has one
+function getGoConfig(content: string): {
+  toolchain?: string;
+  minialGoVersion: string;
+  go: string;
+} {
   const toolchainMatch = regEx(/^toolchain\s*go(?<gover>\d+\.\d+\.\d+)$/m).exec(
     content,
   );
   const toolchainVer = toolchainMatch?.groups?.gover;
-  if (toolchainVer) {
-    logger.debug(
-      `Using go version ${toolchainVer} found in toolchain directive`,
-    );
-    return toolchainVer;
-  }
 
-  // If go.mod doesn't have toolchain directive and has a full go version spec,
-  // for example `go 1.23.6`, pick this version, this doesn't match major.minor version spec.
-  //
-  // This is because when go.mod have same version defined in go directive and toolchain directive,
-  // go will remove toolchain directive from go.mod.
-  //
-  // For example, go will rewrite `go 1.23.5\ntoolchain go1.23.5` to `go 1.23.5` by default,
-  // in this case, the go directive is the toolchain directive.
-  const goFullVersion = regEx(/^go\s*(?<gover>\d+\.\d+\.\d+)$/m).exec(content)
-    ?.groups?.gover;
-  if (goFullVersion) {
-    return goFullVersion;
-  }
+  const goMatch = regEx(/^go\s*(?<gover>\d+(\.\d+)+)$/m).exec(content);
 
-  const re = regEx(/^go\s*(?<gover>\d+\.\d+)$/m);
-  const match = re.exec(content);
-  if (!match?.groups?.gover) {
-    return undefined;
+  // go mod spec says if go directive is missing it's 1.16
+  const goDirective = goMatch?.groups?.gover ?? '1.16';
+  let goVersion = goDirective;
+
+  // partial semver version without patch
+  if (/^\d+\.\d+$/.test(goVersion)) {
+    goVersion = `${goVersion}.0`;
```

Comment:
why do we need to add the zero? 🤔

Reply:
`semver.satisfies('1.17', '>=1.17.0')` is false

Reply:
the regex here only match `*.*`